File: solution/3_FilterableList/3_FilterableList/FilterableArrayListTest.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using src.lists;

namespace _3_FilterableList
{
    [TestClass]
    public class FilterableArrayListTest
    {
        [TestMethod]
        public void filterFiltertAlleElementenUitLijstDieMatchen() {
		    FilterableList<long> filterableList = new FilterableArrayList<long>();
		    filterableList.Add(1L);
            filterableList.Add(4L);
            filterableList.Add(14L);
            filterableList.Add(15L);
            filterableList.Add(16L);
            filterableList.Add(17L);
		
		    List<long> result = filterableList.filter(new GroterDan15FilterPredicate());
		    Assert.AreEqual(2, result.Count);
            Assert.AreEqual(16L, result[0]);
            Assert.AreEqual(17L, result[1]);
	    }
	
	    [TestMethod]
	    public void erIsGeenConnectieTussenGefilterdeLijstEnOrigineleLijst() {
		    FilterableList<long> filterableList = new FilterableArrayList<long>();
            filterableList.Add(15L);
            filterableList.Add(16L);
            filterableList.Add(17L);
		
		    List<long> result = filterableList.filter(new GroterDan15FilterPredicate());

            filterableList.Add(18L);

            Assert.AreEqual(2, result.Count);
            Assert.AreEqual(16L, result[0]);
            Assert.AreEqual(17L, result[1]);
	    }

	    private class GroterDan15FilterPredicate : FilterPredicate<long> {

		    public bool matches(long aLong) {
			    return aLong > 15L;
		    }
		
	    }
    }
}

File: workspace/3_FilterableList/test/lists/FilterableArrayListTest.java
package lists;

import static junit.framework.Assert.assertEquals;

import java.util.List;

import org.junit.Test;

public class FilterableArrayListTest {
	
	@Test
	public void filterFiltertAlleElementenUitLijstDieMatchen() {
		FilterableList<Long> filterableList = new FilterableArrayList<Long>();
		filterableList.add(1L);
		filterableList.add(4L);
		filterableList.add(14L);
		filterableList.add(15L);
		filterableList.add(16L);
		filterableList.add(17L);
		
		List<Long> result = filterableList.filter(new GroterDan15FilterPredicate());
		assertEquals(2, result.size());
		assertEquals((Long)16L, result.get(0));
		assertEquals((Long)17L, result.get(1));
	}
	
	@Test
	public void erIsGeenConnectieTussenGefilterdeLijstEnOrigineleLijst() {
		FilterableList<Long> filterableList = new FilterableArrayList<Long>();
		filterableList.add(15L);
		filterableList.add(16L);
		filterableList.add(17L);
		
		List<Long> result = filterableList.filter(new GroterDan15FilterPredicate());
		
		filterableList.add(18L);
		
		assertEquals(2, result.size());
		assertEquals((Long)16L, result.get(0));
		assertEquals((Long)17L, result.get(1));
	}

	private class GroterDan15FilterPredicate implements FilterPredicate<Long> {

		@Override
		public boolean matches(Long aLong) {
			return aLong.longValue() > 15L;
		}
		
	}
}
